domain/src/main/java/org/sagebionetworks/research/domain/inject/ClassInfo.java

```diff
@@ -30,17 +30,44 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
-package org.sagebionetworks.research.domain.form;
-
-import android.support.annotation.StringDef;
-import java.lang.annotation.Retention;
-import java.lang.annotation.RetentionPolicy;
-
-@Retention(RetentionPolicy.SOURCE)
-@StringDef({InputDataType.BOOLEAN, InputDataType.DATE, InputDataType.DATE_COMPONENTS})
-public @interface InputDataType {
-    String BOOLEAN = "boolean";
-    String DATE = "date";
-    String DATE_COMPONENTS = "dateComponents";
-    // TODO: populate remaining types
+package org.sagebionetworks.research.domain.inject;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.google.gson.JsonDeserializer;
+
+public class ClassInfo<T> {
+    @NonNull
+    private final Class<T> implementationClass;
+    @Nullable
+    private final String typeKey;
+    @Nullable
+    private final JsonDeserializer<? extends T> jsonDeserializer;
+
+    public ClassInfo(@NonNull final Class<T> implementationClass,
```

Comment:
I like that this class reduces the number of providers in Dagger modules. However, I find this class confusing because it is combining concerns and has two optional parameters.
* Mapping a Type to a sub-Type for polymorphic deserialization
* Mapping an interface to a concrete class
* Mapping a class to a JsonDeserializer when Gson's default behavior is insufficient

This ends up with one Map<Class, ClassInfo> being needed as inputs to multiple @Provide methods which use a subset of ClassInfo's fields